src/core/service.c

```diff
@@ -5035,6 +5100,134 @@ static int service_can_clean(Unit *u, ExecCleanMask *ret) {
+
+        /* Given we are running from PID1, avoid doing potentially heavy I/O operations like opening images
+         * directly, and instead fork a worker process. We record the D-Bus message, so that we can reply
+         * after the operation has finished. This way callers can wait on the message and know that the new
+         * resource is available (or the operation failed) once they receive the response. */
+        r = unit_fork_helper_process(u, "(sd-mount-in-ns)", /* into_cgroup= */ false, &worker);
+        if (r < 0) {
+                log_unit_warning_errno(
+                                u,
+                                r,
+                                "Failed to fork process to mount '%s' on '%s' in unit's namespace: %m",
+                                src,
+                                dst);
+                goto fail;
+        }
+        if (r == 0) {
+                if (flags & MOUNT_IN_NAMESPACE_IS_IMAGE)
+                        r = mount_image_in_namespace(
+                                        unit_pid,
+                                        propagate_directory,
+                                        "/run/systemd/incoming/",
+                                        src, dst,
+                                        flags,
+                                        options,
+                                        s->exec_context.mount_image_policy ?: &image_policy_service);
+                else
+                        r = bind_mount_in_namespace(
+                                        unit_pid,
+                                        propagate_directory,
+                                        "/run/systemd/incoming/",
+                                        src, dst,
+                                        flags);
+                if (r < 0)
+                        log_unit_warning_errno(
+                                        u,
+                                        r,
+                                        "Failed to mount '%s' on '%s' in unit's namespace: %m",
+                                        src,
+                                        dst);
+                else
+                        log_unit_debug(u, "Mounted '%s' on '%s' in unit's namespace", src, dst);
+                exit(r < 0 ? EXIT_FAILURE : EXIT_SUCCESS);
+        }
+
+        r = unit_watch_pidref(u, &worker, /* exclusive= */ true);
+        if (r < 0)
+                return r;
+
+        sd_bus_message_unref(s->mount_request); /* Just in case */
+        s->mount_request = sd_bus_message_ref(message);
+        s->control_pid = TAKE_PIDREF(worker);
+        service_set_state(s, SERVICE_MOUNTING);
+        return 0;
+
+fail:
+        s->mount_result = SERVICE_FAILURE_RESOURCES;
+        s->timer_event_source = sd_event_source_disable_unref(s->timer_event_source);
+        return r;
+}
+
+static bool service_can_live_mount(Unit *u, const char *dst, char **error) {
```

Comment:
hmm, thinking about this, i think it would make more sense if the 3rd arg would actually be an sd_bus_error we regularly initialize with to a dbus error + msg.